File: Python/Medium/Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        
        dictionary = {}
        max_length = 0
        i1 = 0
        i2 = 0
        
        while i2 < len(s):
            ch = str(s[i2])
            
            while dictionary.get(ch) is not None:
                ch_to_remove = s[i1]
                dictionary.pop(ch_to_remove)
                i1 += 1
                
            max_length = max(i2 - i1 + 1, max_length)
            
            dictionary[ch] = True
            i2 += 1
        
        
        return max_length
        
        
        """
        :type s: str
        :rtype: int
        """
        

File: Java/Medium/Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        //Time: O(n)
        //Space: O(n)
        HashMap<Character, Integer> hm = new HashMap<Character, Integer>();
        
        int max = 0;
        int start = 0, end = 0;
        while(end < s.length()){
            
            char currChar = s.charAt(end);
            
            if(!hm.containsKey(currChar)){
                max = Math.max(max, end + 1 - start);
            }
            else{
                int prevCharIndex = hm.get(currChar);
                if(prevCharIndex < start){
                    max = Math.max(max, end + 1 - start);
                }
                else start = prevCharIndex + 1;
            }
            
            
            hm.put(currChar, end);
            end++;
        }
              
        return max;
    }
}
